### ISSUE TITLE
speedup fp16

### ISSUE THREAD

**Issue by rydenisbak** [open-mmlab/mmdetection#2131]
You do not need this line.
https://github.com/open-mmlab/mmdetection/blob/master/mmdet/core/fp16/hooks.py#L100

because if input is half tensor, weight is float tensor, output anyway is half tesnor.
But if I comment this line I have speed up near 20% on 1060 and near 50% on 2080.

Check it out please.

**Comment by rydenisbak** [comment]
I checked it on single stage retina-like detector. All worked right.

**Comment by yhcao6** [comment]
Thanks for pointing out that! 
In early versions of pytorch, such as 1.0, 1.1, it is necessary to explicitly convert the half input to float. Otherwise, it will report an error. The error disappears when upgrading pytorch to 1.3.

Would you like to create a pr? We need to compare the version of pytorch to decide if we need to comment this line. I can help you benchmark.

**Comment by clw5180** [comment]
> Thanks for pointing out that!
> In early versions of pytorch, such as 1.0, 1.1, it is necessary to explicitly convert the half input to float. Otherwise, it will report an error. The error disappears when upgrading pytorch to 1.3.
> 
> Would you like to create a pr? We need to compare the version of pytorch to decide if we need to comment this line. I can help you benchmark.

I try to comment '`module.float()`' in Pytorch 1.3, but also got error '**Expected tensor for argument #1 'input' to have the same type as tensor for argument #2 'weight'; but type torch.cuda.FloatTensor does not equal torch.cuda.HalfTensor (while checking arguments for cudnn_batch_norm)**'  how to deal with that ? @yhcao6

**Comment by rydenisbak** [comment]
> Thanks for pointing out that!
> In early versions of pytorch, such as 1.0, 1.1, it is necessary to explicitly convert the half input to float. Otherwise, it will report an error. The error disappears when upgrading pytorch to 1.3.
> 
> Would you like to create a pr? We need to compare the version of pytorch to decide if we need to comment this line. I can help you benchmark.

strange, it seemed to me that I tried this on pytorch 1.0 and 1.1 and 1.2 in other my project. I can check it today in several hours

**Comment by clw5180** [comment]
> Thanks for pointing out that!
> In early versions of pytorch, such as 1.0, 1.1, it is necessary to explicitly convert the half input to float. Otherwise, it will report an error. The error disappears when upgrading pytorch to 1.3.
> 
> Would you like to create a pr? We need to compare the version of pytorch to decide if we need to comment this line. I can help you benchmark.

or do you mean to comment '`module.forward = patch_forward_method(module.forward, torch.half, torch.float)`' ？I test this don't have bug in Pytorch1.1 and 1.3 @yhcao6

**Comment by rydenisbak** [comment]
> > Thanks for pointing out that!
> > In early versions of pytorch, such as 1.0, 1.1, it is necessary to explicitly convert the half input to float. Otherwise, it will report an error. The error disappears when upgrading pytorch to 1.3.
> > Would you like to create a pr? We need to compare the version of pytorch to decide if we need to comment this line. I can help you benchmark.
> 
> I try to comment '`module.float()`' in Pytorch 1.3, but also got error '**Expected tensor for argument #1 'input' to have the same type as tensor for argument #2 'weight'; but type torch.cuda.FloatTensor does not equal torch.cuda.HalfTensor (while checking arguments for cudnn_batch_norm)**' how to deal with that ? @yhcao6

I do not need comment this line, you need to comment line after that

**Comment by clw5180** [comment]
Yes, and it worked for me both Pytorch1.1 and 1.3 @rydenisbak ! How do you find this method? Do you know other method to speed up the fp16?

**Comment by rydenisbak** [comment]
Other method that I tried is do not create a list of fp32 weights during each iteration like this
https://github.com/open-mmlab/mmdetection/blob/master/mmdet/core/fp16/hooks.py#L67

You can create this list once during initialization. But I got small acceleration

**Comment by rydenisbak** [comment]
I noticed that group norm need to be casted into fp32 and then to fp16.
link to the PR
https://github.com/open-mmlab/mmdetection/pull/2151

**Comment by yhcao6** [comment]
@rydenisbak 
It is weird, I set up a new Conda environment, the PyTorch version is 
```
pytorch                   1.1.0           py3.7_cuda10.0.130_cudnn7.5.1_0
```
Then I clone your repo and run [fp16 faster rcnn r-50](https://github.com/open-mmlab/mmdetection/blob/master/configs/fp16/faster_rcnn_r50_fpn_fp16_1x.py), the program fails. 
Here is the training log [train.txt](https://github.com/open-mmlab/mmdetection/files/4250198/train.txt).
If it is convenient, could you show me the **training log** and **detailed PyTorch version** like above?

**Comment by rydenisbak** [comment]
@yhcao6 ok, I will try set up new conda env like above.

Can you try decrease loss scale, for example to 16? 
fp16 = dict(loss_scale=16.)

May be problem disappears

**Comment by rydenisbak** [comment]
@yhcao6 I got the same error. Strange, but if i use singe-stage detector all works right. I will try to solve this problem.

**Comment by rydenisbak** [comment]
I found out what the problem we faced.
In pytorch 1.2 if batchnorm in eval mode it will be cause our problem.
For example, if you change backbone config like this:
```python
backbone=dict(
    type='ResNet',
    depth=50,
    num_stages=4,
    out_indices=(0, 1, 2, 3),
    frozen_stages=-1,
    norm_eval=False, # by default it is True
    style='pytorch'),
```
error disappear.

In pytorch 1.3.1 error disappear anyway.

Unfortunately I have no idea how to solve this problem, now. I need time for thinking

**Comment by rydenisbak** [comment]
@yhcao6, may be we apply changes if pytorch version is 1.3 or above?
```python
def patch_norm_fp32(module):
    if isinstance(module, (nn.modules.batchnorm._BatchNorm, nn.GroupNorm)):
        module.float()
        path_forward_condition = isinstance(module, nn.GroupNorm)
        major, minor, _ = [int(v) for v in torch.__version__.split('.')]
        path_forward_condition |= major < 1 or major == 1 and minor < 3
        if path_forward_condition:
            module.forward = patch_forward_method(module.forward, torch.half,
                                                  torch.float)

    for child in module.children():
        patch_norm_fp32(child)
    return module
```

**Comment by yhcao6** [comment]
Thanks for you hard working! I agree with you. I can help you benchmark if you have no free machines. I will inform you if I have new progress.

**Comment by rydenisbak** [comment]
Thank you, if you can benchmark it will be good! 

Now I have free only GeForce GTX 1060 6Gb. I'm not sure that is good gpu for benchmark. On this gpu I saw ~20% speed up on my own single stage detector. (backbone ~ mobilenet v2 * 0.5)

**Comment by yhcao6** [comment]
Hi,@rydenisbak, this is the result
| Model             | Pytorch | Train time (s/iter) | AP    | Training log |
|-------------------|---------|---------------------|-------|--------------|
| Faster R-CNN R-50 | 1.1     | 0.303               | 0.363 | [comment.log ](https://github.com/open-mmlab/mmdetection/files/4284763/20200303_141203.log) |
| Faster R-CNN R-50 | 1.3     | 0.293               | 0.363 | [no_comment.log](https://github.com/open-mmlab/mmdetection/files/4284770/20200303_163453.log)|

**Comment by rydenisbak** [comment]
Only 3.3%, sad. May be on small networks it will give more effect. In other hand it is faster than before :-)
